File: src/codes/Users_Manual/advanced_examples/custom_gain_0.rs
# extern crate autd3;
# extern crate autd3_driver;
use std::collections::HashMap;
use autd3::prelude::*;
use autd3_driver::derive::*;

#[derive(Gain)]
pub struct FocalPoint {
    position: Vector3,
}

impl FocalPoint {
    pub fn new(position: Vector3) -> Self {
        Self { position }
    }
}

impl Gain for FocalPoint {
    fn calc(&self, geometry: &Geometry, filter: GainFilter) -> Result<HashMap<usize, Vec<Drive>>, AUTDInternalError> {
        Ok(Self::transform(geometry, filter, |dev, tr| Drive::new(
            Phase::from_rad((tr.position() - self.position).norm() * tr.wavelength(dev.sound_speed)),
            EmitIntensity::MAX,
        )))
    }
}
# fn main() { 
# }
#

File: src/codes/Users_Manual/advanced_examples/custom_gain_0.cpp
//~#include "autd3.hpp"
//~#include <unordered_map>
//~#include <limits>
class FocalPoint final : public autd3::Gain<FocalPoint> {
 public:
  explicit FocalPoint(autd3::Vector3 point) : _point(std::move(point)) {}

  std::unordered_map<size_t, std::vector<autd3::Drive>> calc(
      const autd3::Geometry& geometry) const override {
    return transform(geometry, [&](const auto& dev) {
      return [&](const auto& tr) {
        const auto phase = (tr.position() - _point).norm() * dev.wavelength();
        return autd3::Drive{autd3::Phase(phase * autd3::rad),
                            std::numeric_limits<autd3::EmitIntensity>::max()};
      };
    });
  }

 private:
  autd3::Vector3 _point;
};
//~int main() { return 0; }